Fix #4879: Add NPS survey models

adhiamboperes opened the pull request:
## Explanation
Fix #4879.
This is Part 1 of 7 PRs that the add the new survey feature.

This PR adds the proto model and BUILD definitions required for the feature.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

adhiamboperes committed:
Create survey related models

adhiamboperes committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into develop

adhiamboperes committed:
Fix protobuf formatting errors

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-models

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+message CompletedQuestion {
+  // The SurveyUserAnswer given to complete the question.
+  SurveyUserAnswer user_answer = 1;
+}
```
Nit: please add EOF newline.

adhiamboperes replied:
Done

BenHenning commented on model/src/main/proto/survey.proto:
Are you also going to add the BUILD declarations for this proto in this PR, too?

adhiamboperes replied:
Yes, I'll have this in the next update of the PR when I get back.

adhiamboperes replied:
This is done

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+option java_multiple_files = true;
+
+// Corresponds to a single survey.
+message Survey{
```
Nit: space before the ``{`` for consistency. Ditto elsewhere.

adhiamboperes replied:
Done

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+// Corresponds to the semantic name given to each question.
+enum QuestionName{
+  // The question name is unknown.
+   QUESTION_NAME_UNSPECIFIED = 0;
```
Two nits: extra space before the ``Q``, and missing a newline after the ``;``. Ditto elsewhere..

adhiamboperes replied:
Fixed

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+// The list of types a question can have that is used to determine how it is presented by the UI.
+enum QuestionType{
+  // The type of question is unknown.
+  QUESTION_TYPE_UNSPECIFIED=0;
```
Nit: one space on either side of the ``=``, and ditto elsewhere.

adhiamboperes replied:
Fixed.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+  QuestionType question_type = 2;
+
+  // The actual text value of the question.
+  string  question = 3;
```
Nit: an extra space after repeated--please remove. Ditto elsewhere.

adhiamboperes replied:
Fixed

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+  NOT_DISAPPONTED = 3;
+
+  // Corresponds to the answer for  N/A - Don't use Oppia Anymore.
+  NA_DONT_USE_OPPIA_ANYMORE = 4;
```
'NA' is a bit ambiguous--perhaps 'NOT_APPLICABLE' would be clearer?

Also, wouldn't the context be "won't use Oppia anymore"? If they aren't using the app, they presumably wouldn't be filling out this survey.

adhiamboperes replied:
You're right about the context.
I was hoping to avoid extremely long names, but `NOT_APPLICABLE_WONT_USE_OPPIA_ANYMORE` is self documenting.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+  // The question name is unknown.
+   QUESTION_NAME_UNSPECIFIED = 0;
+  // Corresponds to the name of the user type question.
+  QUESTION_USER_TYPE = 1;
```
For this & other proto enums, we don't need to prefix-qualify each one since they'll be part of a generated Java enum. I suggest removing the ``QUESTTION_`` prefix from these constants (except the first one since the unspecified constant should always repeat the enum name per proto convention).

adhiamboperes replied:
Noted. I have fixed this.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+  QuestionName question_name = 4;
+
+  // A list of multi-choice answer options for a given question.
+  repeated  SurveyQuestionAnswer answer_list = 5;
```
How is this populated for question types other than multiple choice?

I'm wondering if using a oneof might be a cleaner way to model these question answers.

adhiamboperes replied:
It is my understanding that the `answer_list` field can be left empty, as the respondent will provide their answer as free-form text.

It is not clear to me why a oneof _answer_ would be needed, since I have modelled the question around QuestionType.  Could you be suggesting that instead of modelling around QuestionType, I might consider using oneof _answer_ ?

BenHenning replied:
One possible issue is that an empty proto value (whether it be text or answer_list) is ambiguous--does it mean "this isn't a multiple choice" or "there aren't any choices defined for this multiple choice question"? It makes it harder to reason about a model if it can take different forms but share properties that should be ignored for certain types (this can result in extra unnecessary error checking needing to be added, or potential for conceptual mix-ups later on).

In this case, you could replace QuestionType with a ``oneof`` and better separate the properties relevant to each type of question, e.g.:

```proto
message SurveyQuestion {
  string question_id = 1;
  QuestionName question_name = 2;
  oneof question_type {
    SurveyQuestionOptionList multiple_choice = 3;
    SurveyQuestionOptionList multiple_choice_with_other = 4;
    bool free_form_text = 5;
  }
}

message SurveyQuestionOptionList {
  repeated SurveyQuestionOption options = 1;
}
```

(Note the extra ``SurveyQuestionOptionList`` is needed because proto doesn't currently support having repeated values in ``oneof``s).

Note how in the case of free-form text we now no longer have to even question what to do with the multiple choice options as a field because the field can't exist in that case.

adhiamboperes replied:
Your proposed `SurveyQuestion` message structure using `oneof` looks good. In this case, I can create a seperate UiModel for QuestionType which will be used for rendering.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+  // The actual question the user is currently viewing.
+  SurveyQuestion  question = 1;
+
+  //Indicates whether this question has a previous question that has been answered,
```
Nit: Exactly one space should be between ``//`` and the first letter. Ditto elsewhere.

adhiamboperes replied:
Corrected

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+}
+
+// Corresponds to a question that the user has previously completed.
+message CompletedQuestion {
```
Will this ever contain anything other than ``SurveyUserAnswer``? If not, we could just reference that directly.

adhiamboperes replied:
At the moment, that's all it contains, so I have  referenced it directly.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+    // A previous question completed by the user.
+    CompletedQuestion completed_question = 5;
+
+    // True if this is the last question in the survey.
```
Probably more like: "Present if this is the last question in the survey (the actual value of the boolean does not matter)."

We may set it to true, but just its presence as being set is the key indicator (per how we handle this with other bool oneof cases).

adhiamboperes replied:
Refactored the comment.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+// once the user finishes or exits the survey session.
+message EphemeralSurveyQuestion{
+  // The actual question the user is currently viewing.
+  SurveyQuestion  question = 1;
```
Nit: one space between the type & name.

adhiamboperes replied:
Fixed

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+  SURVEY_MANDATORY_PART_COMPLETED = 3;
+
+  // Corresponds to the optional part of the survey being completed.
+  SURVEY_COMPLETED = 4;
```
Perhaps 'FULLY_COMPLETED' might better correspond to 'MANDATORY_PART_COMPLETED'?

adhiamboperes replied:
This sounds good.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+
+// Corresponds to a single survey.
+message Survey{
+  // The ID of the survey.
```
For this & other IDs, we should try to explain what the ID format is or if it's an opaque type provided by the server (which I don't think it is in this case), and its uniqueness guarantees.

Technically, our audience for these doc comments are both callers to set and read the values, but I personally find it more helpful to focus on the proto's contract on the "read" side and assume the setter will comply with it.

adhiamboperes replied:
I have updated the documentation for the IDs. Please let me know if this covers everything.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+
+// Represents the multiple choice answer available for a question.
+message SurveyQuestionAnswer{
+  // The ID of this answer option.
+  string question_answer_id = 1;
+
+  // The semantic name of the question this answer option is tied to.
+  QuestionName question_name = 2;
```
Why do these values need to be replicated for only a multiple choice option?

adhiamboperes replied:
This message represent the answer in general, and according to the current implementation, is empty when the question is not a multiple choice question.

BenHenning replied:
I think this relates to the oneof case above, so I'll respond there.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+}
+
+// Represents the multiple choice answer available for a question.
+message SurveyQuestionAnswer{
```
I don't think this name well conveys that it represents a multiple choice option--I suggest renaming.

adhiamboperes replied:
Renamed to `SurveyAnswerOption`

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,182 @@
+  QuestionName question_name = 2;
+
+  // The value of the answer the user gave.
+  string user_answer = 3;
```
What is this value for multiple choice--the actual label text? If so, is that what we're supposed to be submitting to the backend (vs. an index or something)? If so, how will this work with translated versions of the multiple choice options?

adhiamboperes replied:
Yes, it was meant to be the actual label text.
I see your point about translated text, and I think it might be advisable to use the associated enums instead.

BenHenning replied:
Enums sound good if we always have a fixed set of text choices to choose from.

adhiamboperes replied:
At the moment, all the MCQs have a fixed set of answers, so I have opted to create a oneof of each of those answer representations that this field could take.

adhiamboperes replied:
Re: translations, I have added a language code field to the question so that we can query open-ended responses based on this. It occurred to me that a generic query will just show all the answers without the language context.

BenHenning replied:
Given the change, is this field still needed? I'm assuming the question type provides enough context to allow the UI to pick app strings to show for each question, so it isn't clear to me why this string will be needed.

adhiamboperes replied:
It is no-longer needed. Please counter-check, I removed it in a previous commit.

BenHenning reviewed (changes requested):
Thanks @adhiamboperes! Took a first pass--PTAL.

adhiamboperes committed:
Fix formating nits

adhiamboperes committed:
Refactor MarketFitAnswer field   NA_DONT_USE_OPPIA_ANYMORE to NOT_APPLICABLE_WONT_USE_OPPIA_ANYMORE

'NA' is a bit ambiguous, and the context is "won't use Oppia anymore". If they aren't using the app, they presumably wouldn't be filling out this survey.

adhiamboperes committed:
Refactor enum constants to not include the enum name.

For proto enums, we don't need to prefix-qualify each one since they'll be part of a generated Java enum -- except the first one since the unspecified constant should always repeat the enum name per proto convention.

adhiamboperes committed:
Refactor message ID documentation

Refactored to explain the ID format and the uniqueness guaranteed.

adhiamboperes committed:
Refactor SurveyQuestionAnswer to SurveyAnswerOption

To convey that it is a multiple choice answer option.

adhiamboperes reviewed (commented):
Hey @BenHenning, aside from these two review comments, I have addressed all of your previous comments and I'd like you to take another look. Thanks.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -47,31 +51,33 @@ enum QuestionType{
-message SurveyQuestionAnswer{
-  // The ID of this answer option.
+message SurveyAnswerOption {
+  // The ID of this answer option. It is a string value that uniquely identifies the answer option
```
Is it unique to the survey instance, profile, device installation, or global? Ditto elsewhere for new properties where we need to guarantee some level of uniqueness.

BenHenning replied:
This was addressed in the latest changes.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -163,20 +171,15 @@ message EphemeralSurveyQuestion{
-    // A previous question completed by the user.
-    CompletedQuestion completed_question = 5;
+    // The SurveyUserAnswer given to complete the previous question.
+    SurveyUserAnswer user_answer = 5;
```
```suggestion
    SurveyUserAnswer completion_user_answer = 5;
```

Perhaps to keep the previous context that this represents a "completed" state?

adhiamboperes replied:
I took this suggestion

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -6,36 +6,40 @@ option java_package = "org.oppia.android.app.model";
-  // The ID of the survey.
+message Survey {
+  // The ID of the survey. This is a string value that uniquely identifies the survey.
+  // The ID is guaranteed to be unique across all surveys.
```
Unique across all surveys globally, for the device installation, or the profile?

adhiamboperes replied:
SurveyIds are unique globally, since that is what we will use to group the responses given by a unique profile - if needed.

BenHenning reviewed (commented):
Thanks @adhiamboperes! Took an incremental pass. PTAL.

Once all of these are addressed, I'll take one more full pass just to make sure I didn't miss anything.

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-models

adhiamboperes committed:
Refactor SurveyQuestion to use oneof instead of repeated answer list

Leaving a proto list empty raises the potential ambiguity of an empty proto value, which can make it harder to reason about the model.

By using oneof, we can better control the structure of the message and make it clearer what fields are relevant to each question type.

adhiamboperes committed:
Refactor SurveyQuestion to have a language code

We can use this field to filter open-ended responses in analytics

adhiamboperes committed:
Add BUILD definitions for survey.proto

adhiamboperes committed:
Refactor SurveyUserAnswer to contain a oneof of the answer enums

adhiamboperes committed:
Refactored messages for readability and consistency.

adhiamboperes committed:
Remove unused import in oppia_logger.proto

adhiamboperes committed:
Rename enum field causing conflicts

STARTED_NOT_COMPLETED is already used in topic.

BenHenning commented on model/src/main/proto/oppia_logger.proto:
```diff
@@ -2,7 +2,6 @@ syntax = "proto3";
 
 package model;
 
-import "exploration.proto";
```
Ah, I was removing this in one of my chained PRs since I also noticed the warning--thanks for removing it. :)

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -61,43 +60,64 @@ enum QuestionName {
+    MarketFitAnswer market_fit = 4;
+
+    // The integer value representing the nps score selected by the user.
+    int32 nps = 5;
```
Perhaps ``nps_score``, instead, per the comment? Ditto above.

adhiamboperes replied:
I've updated both fields.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,204 @@
+// Corresponds to a single survey.
+message Survey {
+  // The ID of the survey. This is a string value that uniquely identifies the survey.
+  // The ID is guaranteed to be unique across all surveys globally for the purpose of analysis.
```
```suggestion
  // The ID should ostensibly be unique across all surveys globally for the purpose of analysis.
```

We don't actually make guarantees since that would require a remote service like Firestore keeping track of used IDs. GUIDs are statistically unlikely to overlap, but that isn't the same as a guarantee.

adhiamboperes replied:
Thanks, this makes sense. I can see why we really can't make the guarantee when thousands of devices are involved.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,204 @@
+
+// Represents the different types of users who could be taking the survey.
+// Consists of answer options for the user type question.
+enum  UserTypeAnswer {
```
```suggestion
enum UserTypeAnswer {
```

adhiamboperes replied:
Fixed

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,204 @@
+
+// Represents general opinion of the user about the Oppia android app.
+// Consists of answer options for the market fit question.
+enum  MarketFitAnswer {
```
```suggestion
enum MarketFitAnswer {
```

adhiamboperes replied:
Fixed.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,204 @@
+  // Indicates whether this question has an upcoming question that can be navigated to.
+  bool has_next_question = 3;
+
+  //  Indicates possible states a question can take depending on whether a user needs to provide an
+  //  answer or has navigated to a previous question, or has reached the last question.
```
```suggestion
  // Indicates possible states a question can take depending on whether a user needs to provide an
  // answer or has navigated to a previous question, or has reached the last question.
```

adhiamboperes replied:
Fixed.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,204 @@
+  // User type unknown.
+  USER_TYPE_UNSPECIFIED = 0;
+
+  // Corresponds to an unspecified or unknown type of user.
```
Out of curiosity, why not use ``USER_TYPE_UNSPECIFIED`` for this case? Why have a different OTHER enum?

adhiamboperes replied:
Loosely speaking, it would be more difficult to distinguish between responses that were genuinely unspecified and those that were intended to be "other" choices. However the implementation should have checks to ensure each question has a selected answer before proceeding to the next question.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,204 @@
+}
+
+// Corresponds to the semantic name given to each question.
+enum QuestionName {
```
Should this be ``SurveyQuestionName`` for consistency elsewhere in this file? It seems like everything tends to start with 'survey'.

adhiamboperes replied:
> It seems like everything tends to start with 'survey'.

This is only because some of the message names have been used in other messages. 

I have updated this message to start with survey too, for consistency.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,204 @@
+}
+
+// Represents an answer that was selected by a user for a survey question.
+message SurveyUserAnswer {
```
Should this maybe be ``SurveyQuestionAnswer``? The 'User' part is throwing me off--I think this is meant to have parity with ``SurveyQuestionOption`` above, so a rename might help connect them better.

adhiamboperes replied:
I'm inclined to go with `SurveySelectedAnswer` so that it can make sense also in the context of `EphemeralSurveyQuestion`

BenHenning replied:
That sounds good--thanks!

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,204 @@
+  }
+
+  // The language code to which this question is localized.
+  string language_code = 7;
```
We should use OppiaLanguage for this purpose, see:

https://github.com/oppia/oppia-android/blob/1bbadac452d87fa7748bcb5d5ad3a8b446473848/model/src/main/proto/languages.proto#L9

adhiamboperes replied:
Done

BenHenning reviewed (commented):
Thanks @adhiamboperes! Took another pass--PTAL.

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-models

adhiamboperes committed:
Refactor field names

Also fixed some spacing issues and updated comments.

BenHenning commented on model/src/main/proto/survey.proto:
```diff
@@ -0,0 +1,206 @@
+  // The semantic name of the question this answer option is tied to.
+  SurveyQuestionName question_name = 2;
+
+  // The actual text value of the answer option.
+  string label = 3;
```
This part still seems wrong, unless I'm missing something. Since we localize text in the frontend, should this be an enum, instead?

Ditto for 'question' in SurveyQuestion above.

adhiamboperes replied:
I have removed both string fields since we already have `SurveyQuestionName` in SurveyQuestion and `oneof answer`  in SurveyQuestionOption.

BenHenning reviewed (commented):
Thanks @adhiamboperes! The proto is looking really good, just had one comment about two fields left--PTAL.

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into nps-survey-models

adhiamboperes committed:
Remove string fields in SurveyQuestion and SurveyQuestionOption

Since we localize text in the frontend, these are represented by enums; SurveyQuestionName and oneof answer respectively.

adhiamboperes committed:
Fix skipped field numbers

Reorder field numbers after removing some fields.

BenHenning reviewed (approved):
Thanks @adhiamboperes. This LGTM!

I'll leave it to you to merge since it'll affect your downstream chain PR.

adhiamboperes committed:
Merge branch 'develop' into nps-survey-models
